File: node-depths/main.iterative.ts
class BinaryTree {
  value: number;
  left: BinaryTree | null;
  right: BinaryTree | null;

  constructor(value: number) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}

type StackItem = {
  node: BinaryTree | null;
  depth: number;
};

export function nodeDepths(root: BinaryTree) {
  let sumOfDepths = 0;
  let stack: StackItem[] = [{ node: root, depth: 0 }];

  while (stack.length) {
    let nodeInfo = stack.pop();

    if (!nodeInfo) {
      continue;
    }

    let { node, depth } = nodeInfo;

    if (!node) {
      continue;
    }

    sumOfDepths += depth;

    stack.push({ node: node.left, depth: depth + 1 });
    stack.push({ node: node.right, depth: depth + 1 });
  }

  return sumOfDepths;
}

export {};

File: node-depths/main.iterative.py
class BinaryTree:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


def nodeDepths(root):
    # initialise the sum of the depths
    sumOfDepths = 0;
    # create a stack with each node and its corresponding depth, ready to be processed
    stack = [{ "node": root, "depth": 0}]
	
    # while there are items in the stack
    while len(stack) > 0:
        # grab a node from the stack
        nodeInfo = stack.pop()
        node, depth = nodeInfo["node"], nodeInfo["depth"]
		
        if node is None:
            continue

        # increase the depths sum by the depth of the node
        sumOfDepths += depth

        # append the node's children to the stack, incrementing their depth count by 1
        stack.append({ "node": node.left, "depth": depth + 1})
        stack.append({ "node": node.right, "depth": depth + 1})
	
    # once the stack is empty, return the sum of the depths
    return sumOfDepths


